entity-service-impl/src/main/java/org/hypertrace/entity/fetcher/EntityFetcher.java

```diff
@@ -27,34 +30,44 @@ public EntityFetcher(
     this.parser = documentParser;
   }
 
-  public List<Entity> getEntitiesByDocIds(String tenantId, Collection<String> docIds) {
-    return query(buildExistingEntitiesByDocIdQuery(tenantId, docIds))
-        .collect(Collectors.toUnmodifiableList());
+  public List<Entity> getEntitiesByDocIds(String tenantId, Collection<String> docIds)
+      throws IOException {
+    return query(buildExistingEntitiesByDocIdQuery(tenantId, docIds)).collect(toUnmodifiableList());
   }
 
   public List<Entity> getEntitiesByEntityIds(
-      String tenantId, java.util.Collection<String> entityIds) {
+      String tenantId, java.util.Collection<String> entityIds) throws IOException {
     if (entityIds.isEmpty()) {
       return Collections.emptyList();
     }
     return query(buildExistingEntitiesByEntityIdQuery(tenantId, entityIds))
-        .collect(Collectors.toUnmodifiableList());
+        .collect(toUnmodifiableList());
   }
 
-  public Stream<Entity> query(org.hypertrace.core.documentstore.Query query) {
-    return Streams.stream(this.entitiesCollection.search(query))
-        .map(this::entityFromDocument)
-        .flatMap(Optional::stream)
-        .map(Entity::toBuilder)
-        .map(Entity.Builder::build);
+  public Stream<Entity> query(org.hypertrace.core.documentstore.Query query) throws IOException {
+    final CloseableIterator<Document> iterator = this.entitiesCollection.search(query);
```

Comment:
So we had to leave it open for the stream. Do we still need the stream or can we convert places to use the list? just .stream() it for compat if needed. If for some reason we want full compat, I'd suggest having this implemented as just `return this.query(query).stream();` Only potential downside is if there are memory reasons we're trying to avoid holding all entities in memory at one time.

Reply:
> So we had to leave it open for the stream. 

Yeah. There are 3 usages of this method, 2 of them need just the list. The third usage is actually streaming the response to the caller (1 at a time).

```java
try {
      this.entityFetcher
          .query(DocStoreConverter.transform(tenantId.get(), request, Collections.emptyList()))
          .forEach(responseObserver::onNext);
    } catch (final Exception e) {
      LOG.error("Unknown error occurred", e);
      responseObserver.onError(
          Status.INTERNAL
              .withDescription("Unknown error occurred")
              .asRuntimeException(requestContext.buildTrailers()));
    }

    responseObserver.onCompleted();
```
And, that requires us to keep the stream open.

Reply:
In that third case - which is itself ridiculous - we can do the same with the list and close up the iterator without waiting for the client to iterate themselves. We don't want to leave the connection to the collection open at the whim of the caller.